underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSInputStream.java

```diff
@@ -66,23 +73,47 @@
    * @param key the key of the file
    * @param client the client for GCS
    * @param pos the position to start
-   * @throws ServiceException if a service exception occurs
+   * @param retryPolicy retry policy in case the key does not exist
    */
-  GCSInputStream(String bucketName, String key, GoogleStorageService client, long pos)
-      throws ServiceException {
+  GCSInputStream(String bucketName, String key, GoogleStorageService client,
+      long pos, RetryPolicy retryPolicy) throws ServiceException {
     mBucketName = bucketName;
     mKey = key;
     mClient = client;
     mPos = pos;
-    // For an empty file setting start pos = 0 will throw a ServiceException
-    if (mPos > 0) {
-      mObject = mClient.getObject(mBucketName, mKey, null, null, null, null, mPos, null);
-    } else {
-      mObject = mClient.getObject(mBucketName, mKey);
-    }
+    mObject = getObjectWithRetry(retryPolicy);
     mInputStream = new BufferedInputStream(mObject.getDataInputStream());
   }
 
+  /**
+   * Retries getting a {@link GSObject}.
+   *
+   * @param retryPolicy the retry policy to solve eventual consistency issue
+   * @return the {@link GSObject}
+   */
+  private GSObject getObjectWithRetry(RetryPolicy retryPolicy) throws ServiceException {
+    ServiceException lastException = null;
+    while (retryPolicy.attempt()) {
+      try {
+        if (mPos > 0) {
+          return mClient.getObject(mBucketName, mKey, null, null, null, null, mPos, null);
+        } else {
+          return mClient.getObject(mBucketName, mKey);
+        }
+      } catch (ServiceException e) {
+        LOG.warn("Attempt {} to open key {} in bucket {} failed with exception : {}",
+            retryPolicy.getAttemptCount(), mKey, mBucketName, e.toString());
+        if (e.getResponseCode() != HttpStatus.SC_NOT_FOUND) {
+          throw e;
```

Comment:
same here

Reply:
For `GCSInputStream`, the `ServiceException` is directly thrown in the constructor, because 
```
  GCSInputStream(String bucketName, String key, GoogleStorageService client,
      long pos, RetryPolicy retryPolicy) throws ServiceException {
     .....
    mObject = getObjectWithRetry(retryPolicy);
    mInputStream = new BufferedInputStream(mObject.getDataInputStream());
  }
```
both `getObjectWithRetry` and `getDataInputStream` will throw `ServiceException`.

The `ServiceException` is caught and transformed to `IOException` in `GCSUnderFileSystem..openObject`

Reply:
got it, thx